QModManager/API/ModLoading/Internal/QModFactory.cs

```diff
@@ -0,0 +1,211 @@
+
+                    continue;
+                }
+
+                if (!mod.Enable)
+                {
+                    Logger.Info($"Mod \"{mod.DisplayName}\" is disabled via config, skipping...");
+
+                    continue;
+                }
+
+                SortResults result = sortedMods.Add(mod);
+
+                switch (result)
+                {
+                    case SortResults.CircularLoadOrder:
+
+                        break;
+                    case SortResults.CircularDependency:
+
+                        break;
+                    case SortResults.DuplicateId:
+                        // TODO Report error
+                        break;
+                }
+            }
+
+            List<QMod> modsToLoad = sortedMods.CreateFlatValueList();
+
+            this.FailedToCreate = ReportErrors(sortedMods.GetErrors());
+            CheckOldHarmony(modsToLoad);
+
+            return modsToLoad;
+        }
+
+        private static int ReportErrors(List<string> erroredMods)
+        {
+            if (erroredMods.Count == 0)
+                return 0;
+
+            string display = "The following mods could not be loaded: ";
+            int maxDisplay = Math.Min(5, erroredMods.Count);
+
+            for (int i = 0; i < maxDisplay; i++)
+            {
+                display += $", {erroredMods[i]}";
+            }
+
+            if (erroredMods.Count > 5)
+                display += $", and {erroredMods.Count - 5} others";
+
+            display += ". Check the log for details.";
+
+            Dialog.Show(display, Dialog.Button.SeeLog, Dialog.Button.Close, false);
+
+            return erroredMods.Count;
+        }
+
+        private static void CheckOldHarmony(List<QMod> loadedMods)
+        {
+            var modsThatUseOldHarmony = new List<QMod>();
+
+            foreach (QMod mod in loadedMods)
+            {
+                AssemblyName[] references = mod.LoadedAssembly.GetReferencedAssemblies();
+                foreach (AssemblyName reference in references)
+                {
+                    if (reference.FullName == "0Harmony, Version=1.0.9.1, Culture=neutral, PublicKeyToken=null")
+                    {
+                        modsThatUseOldHarmony.Add(mod);
+                        break;
+                    }
+                }
+            }
+
+            if (modsThatUseOldHarmony.Count > 0)
+            {
+                Logger.Warn($"Some mods are using an old version of harmony! This will NOT cause any problems, but it's not recommended:");
+                foreach (IQMod mod in modsThatUseOldHarmony)
+                {
+                    Console.WriteLine($"- {mod.DisplayName} ({mod.Id})");
+                }
+            }
+        }
+
+        private static QMod FromDll(string subDirectory, string[] dllFilePaths)
+        {
+            foreach (string dllFile in dllFilePaths)
+            {
+                var assembly = Assembly.LoadFrom(dllFile);
```

Comment:
@AlexejheroYTB 
Is there potential issue if we load an assembly that wasn't needed?  
ex. An extra DLL included by a mod that it uses.

Reply:
Not really a problem. Although we should probably load it into a new `AppDomain` (`AppDomain.CreateDomain`) and then either load it into the main domain if it's a mod dll or unload it by unloading the domain.

Reply:
This would most likely require adding a [new dependency](https://github.com/jduv/AppDomainToolkit)

Reply:
Heads up. I'm going to need your help for this one.  
Dynamically loading DLLs at runtime isn't something I've worked with before.

Reply:
Important question: do we want to unload unnecessary assemblies after QMod attribute was found?